File: projects/02-cloud-solutions/PRJ-CLOUD-001/src/handlers/create.js
'use strict';

const { randomUUID } = require('crypto');

const log = (level, message, context = {}) => {
  console.log(JSON.stringify({ level, message, ...context }));
};

const response = (statusCode, body) => ({
  statusCode,
  headers: { 'Content-Type': 'application/json' },
  body: JSON.stringify(body),
});

module.exports.handler = async (event) => {
  log('INFO', 'Received create request', { requestId: event?.requestContext?.requestId });

  const timestamp = new Date().toISOString();
  const payload = typeof event.body === 'string' ? JSON.parse(event.body || '{}') : event.body || {};
  const id = randomUUID();

  const item = {
    id,
    name: payload.name || 'unnamed',
    createdAt: timestamp,
    metadata: payload.metadata || {},
  };

  log('DEBUG', 'Prepared item for persistence', { item });

  // Placeholder for DynamoDB persistence call; mocked for portfolio demonstration.
  const saved = { ...item, persisted: true };

  return response(201, { message: 'Item created', item: saved });
};

File: projects/p11-serverless/src/lambda/create.py
#!/usr/bin/env python3
"""Lambda function to create items."""
import json
import os
import boto3
from datetime import datetime

dynamodb = boto3.resource('dynamodb')
table_name = os.environ.get('TABLE_NAME', 'items-table')
table = dynamodb.Table(table_name)

def lambda_handler(event, context):
    """Create a new item."""
    try:
        body = json.loads(event['body'])
        item = {
            'id': str(datetime.now().timestamp()),
            'name': body['name'],
            'description': body.get('description', ''),
            'created_at': datetime.now().isoformat()
        }

        table.put_item(Item=item)

        return {
            'statusCode': 201,
            'headers': {'Content-Type': 'application/json'},
            'body': json.dumps(item)
        }
    except Exception as e:
        return {
            'statusCode': 500,
            'body': json.dumps({'error': str(e)})
        }
